### PR TITLE
feat(variadics): additions to variadics including collection types

### PR DESCRIPTION
**Pull request by jhellerstein**
adds a number of features:

collection types for variadics (sets, multisets) that allow search via RefVars (variadic of refs)
into_option (convert a variadic to a variadic of options)
into_vec (convert a variadic to a variadic of vecs)

### CODE REVIEW

File: variadics/src/lib.rs [hunk 1]
```diff
@@ -125,6 +127,18 @@ pub trait VariadicExt: Variadic {
...
+
+    /// The length of this variadic type
+    fn len(&self) -> usize {
+        Self::LEN
+    }
+
+    /// Checks if this variadic type is empty.
+    fn is_empty(&self) -> bool {
+        Self::LEN == 0
+    }
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:141]: When are these needed? Instead of just `Self::LEN`

File: variadics/src/lib.rs [hunk 2]
```diff
@@ -176,7 +190,18 @@ pub trait VariadicExt: Variadic {
...
+
+    /// type for all elements of the variadic being wrapped in Option
+    type IntoOption;
+    /// wrap all elements of the variadic in Option
+    fn into_option(self) -> Self::IntoOption;
+
+    /// type for all elements of the variadic being wrapped in Vec
+    type IntoVec: VecVariadic<UnVec = Self> + Default;
+    /// wrap all elements of the variadic in a Vec
+    fn into_singleton_vec(self) -> Self::IntoVec;
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:202]: nit:
```rust
/// type for all elements of the variadic being wrapped in Option
```
should be
```rust
/// Type for all elements of the variadic being wrapped in `Option`.
```
etc for comments

File: variadics/src/lib.rs [hunk 3]
```diff
@@ -198,6 +223,14 @@ where
         let (item, rest) = self;
         rest.reverse().extend((item, ()))
     }
+    fn reverse_ref(this: Self::AsRefVar<'_>) -> <Self::Reverse as VariadicExt>::AsRefVar<'_> {
+        let (item, rest) = this;
+        let out = Rest::reverse_ref(rest).extend((item, ()));
+        // TODO!!!
+        let out2 = unsafe { std::mem::transmute_copy(&out) };
+        std::mem::forget(out);
+        out2
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:232]: `// TODO(mingwei)`
Not sure if there is a safe way to do this, but should add a safety comment/convince ourselves it is sane

File: variadics/src/lib.rs [hunk 4]
```diff
@@ -297,7 +350,7 @@ pub trait EitherRefVariadic: Variadic {
     /// let un_ref: <var_type!(&u32, &String, &bool) as EitherRefVariadic>::UnRefVar =
     ///     var_expr!(1_u32, "Hello".to_owned(), false);
     /// ```
-    type UnRefVar: VariadicExt;
+    type UnRefVar: for<'a> VariadicExt;
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:353]: the `for<'a>` seems unused?

File: variadics/src/lib.rs [hunk 5]
```diff
@@ -319,6 +372,9 @@ pub trait EitherRefVariadic: Variadic {
     ///
     /// Conversion from `&` to `&mut` is generally invalid, so a `ref_to_mut()` method does not exist.
     type MutVar: MutVariadic<UnRefVar = Self::UnRefVar, MutVar = Self::MutVar>;
+
+    /// convert entries to <UnRefVar as VariadicExt>::AsRefVar
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:376]:
```rust
/// Convert entries to [`<UnRefVar as VariaidcExt>::AsRefVar`](VariadicExt::AsRefVar).
```

File: variadics/src/lib.rs [hunk 6]
```diff
@@ -438,50 +506,39 @@ impl CopyRefVariadic for () {
     fn copy_var(&self) -> Self::UnRefVar {}
 }
 
-/// Clone a variadic of references [`EitherRefVariadic`] into a variadic of owned values [`EitherRefVariadic::UnRefVar`].
+/// Clone a variadic of references [`AsRefVar`] into a variadic of owned values.
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:509]:
```
[`AsRefVar`](VariadicExt::AsRefVar)
```

File: variadics/src/lib.rs [hunk 7]
```diff
@@ -438,50 +506,39 @@ impl CopyRefVariadic for () {
...
 ///     var_expr!(1, "hello world".to_owned(), vec![1, 2, 3]),
 ///     clone_var
 /// );
 /// ```
 #[sealed]
-pub trait CloneRefVariadic: EitherRefVariadic {
-    /// Clone self per-value.
-    fn clone_var(&self) -> Self::UnRefVar;
+pub trait CloneVariadic: VariadicExt + Clone {
+    /// Clone a variadic of references [`AsRefVar`] into a variadic of owned values.
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:522]: ^

File: variadics/src/lib.rs [hunk 8]
```diff
@@ -588,36 +645,218 @@ where
...
+    where
+        R: std::ops::RangeBounds<usize> + Clone;
+}
+
+#[sealed]
+impl<Item, Rest> VecVariadic for (Vec<Item>, Rest)
+where
+    Rest: VecVariadic,
+    // Item: 'static,
+    // Rest: 'static,
```
Comment (MingweiSamuel) [inline variadics/src/lib.rs:794]: remove

File: variadics/src/variadic_collections.rs [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+
+impl<K> Extend<K> for VariadicColumnMultiset<K>
+where
+    K: Eq + Hash + PartialEqVariadic,
+    for<'a> K::AsRefVar<'a>: Hash,
+{
+    // #[cfg_attr(feature = "inline-more", inline)]
+    fn extend<T: IntoIterator<Item = K>>(&mut self, iter: T) {
+        let iter = iter.into_iter();
+        // self.table.reserve(reserve, hasher);
```
Comment (MingweiSamuel) [inline variadics/src/variadic_collections.rs:582]: remove commented-out code

File: variadics/src/lib.rs (hunk 1 above)
Comment (jhellerstein) [reply variadics/src/lib.rs:141]: `len` just matches convention. and once you define `len`, rust warns you to define `is_empty`. Which suggests that using `len` convention was good hygiene?

File: variadics/src/lib.rs (hunk 3 above)
Comment (jhellerstein) [reply variadics/src/lib.rs]: simply changed comment for now

**Review by MingweiSamuel — APPROVED** [review]

**Pull request merged**